File: go/indicators/custom/goertzelspectrum/estimator.go
package goertzelspectrum

import "math"

// estimator implements the Goertzel spectrum estimator. It is an unexported port
// of MBST's GoertzelSpectrumEstimator used only by the GoertzelSpectrum indicator.
type estimator struct {
	length                          int
	spectrumResolution              int
	lengthSpectrum                  int
	minPeriod                       float64
	maxPeriod                       float64
	isFirstOrder                    bool
	isSpectralDilationCompensation  bool
	isAutomaticGainControl          bool
	automaticGainControlDecayFactor float64

	inputSeries          []float64
	inputSeriesMinusMean []float64
	spectrum             []float64
	period               []float64

	// Pre-computed trigonometric tables.
	frequencySin  []float64 // first-order only
	frequencyCos  []float64 // first-order only
	frequencyCos2 []float64 // second-order only

	mean                float64
	spectrumMin         float64
	spectrumMax         float64
	previousSpectrumMax float64
}

// newEstimator creates a new Goertzel spectrum estimator.
func newEstimator(
	length int,
	minPeriod, maxPeriod float64,
	spectrumResolution int,
	isFirstOrder, isSpectralDilationCompensation, isAutomaticGainControl bool,
	automaticGainControlDecayFactor float64,
) *estimator {
	const twoPi = 2 * math.Pi

	lengthSpectrum := int((maxPeriod-minPeriod)*float64(spectrumResolution)) + 1

	e := &estimator{
		length:                          length,
		spectrumResolution:              spectrumResolution,
		lengthSpectrum:                  lengthSpectrum,
		minPeriod:                       minPeriod,
		maxPeriod:                       maxPeriod,
		isFirstOrder:                    isFirstOrder,
		isSpectralDilationCompensation:  isSpectralDilationCompensation,
		isAutomaticGainControl:          isAutomaticGainControl,
		automaticGainControlDecayFactor: automaticGainControlDecayFactor,
		inputSeries:                     make([]float64, length),
		inputSeriesMinusMean:            make([]float64, length),
		spectrum:                        make([]float64, lengthSpectrum),
		period:                          make([]float64, lengthSpectrum),
	}

	result := float64(spectrumResolution)

	if isFirstOrder {
		e.frequencySin = make([]float64, lengthSpectrum)
		e.frequencyCos = make([]float64, lengthSpectrum)

		for i := 0; i < lengthSpectrum; i++ {
			period := maxPeriod - float64(i)/result
			e.period[i] = period
			theta := twoPi / period
			e.frequencySin[i] = math.Sin(theta)
			e.frequencyCos[i] = math.Cos(theta)
		}
	} else {
		e.frequencyCos2 = make([]float64, lengthSpectrum)

		for i := 0; i < lengthSpectrum; i++ {
			period := maxPeriod - float64(i)/result
			e.period[i] = period
			e.frequencyCos2[i] = 2 * math.Cos(twoPi/period)
		}
	}

	return e
}

// calculate fills mean, inputSeriesMinusMean, spectrum, spectrumMin, spectrumMax
// from the current inputSeries contents.
func (e *estimator) calculate() {
	// Subtract the mean from the input series.
	mean := 0.0
	for i := 0; i < e.length; i++ {
		mean += e.inputSeries[i]
	}

	mean /= float64(e.length)

	for i := 0; i < e.length; i++ {
		e.inputSeriesMinusMean[i] = e.inputSeries[i] - mean
	}

	e.mean = mean

	// Seed with the first bin.
	spectrum := e.goertzelEstimate(0)
	if e.isSpectralDilationCompensation {
		spectrum /= e.period[0]
	}

	e.spectrum[0] = spectrum
	e.spectrumMin = spectrum

	if e.isAutomaticGainControl {
		e.spectrumMax = e.automaticGainControlDecayFactor * e.previousSpectrumMax
		if e.spectrumMax < spectrum {
			e.spectrumMax = spectrum
		}
	} else {
		e.spectrumMax = spectrum
	}

	for i := 1; i < e.lengthSpectrum; i++ {
		spectrum = e.goertzelEstimate(i)
		if e.isSpectralDilationCompensation {
			spectrum /= e.period[i]
		}

		e.spectrum[i] = spectrum

		if e.spectrumMax < spectrum {
			e.spectrumMax = spectrum
		} else if e.spectrumMin > spectrum {
			e.spectrumMin = spectrum
		}
	}

	e.previousSpectrumMax = e.spectrumMax
}

func (e *estimator) goertzelEstimate(j int) float64 {
	if e.isFirstOrder {
		return e.goertzelFirstOrderEstimate(j)
	}

	return e.goertzelSecondOrderEstimate(j)
}

func (e *estimator) goertzelSecondOrderEstimate(j int) float64 {
	cos2 := e.frequencyCos2[j]

	var s1, s2 float64

	for i := 0; i < e.length; i++ {
		s0 := e.inputSeriesMinusMean[i] + cos2*s1 - s2
		s2 = s1
		s1 = s0
	}

	spectrum := s1*s1 + s2*s2 - cos2*s1*s2
	if spectrum < 0 {
		return 0
	}

	return spectrum
}

func (e *estimator) goertzelFirstOrderEstimate(j int) float64 {
	cosTheta := e.frequencyCos[j]
	sinTheta := e.frequencySin[j]

	var yre, yim float64

	for i := 0; i < e.length; i++ {
		re := e.inputSeriesMinusMean[i] + cosTheta*yre - sinTheta*yim
		im := e.inputSeriesMinusMean[i] + cosTheta*yim + sinTheta*yre
		yre = re
		yim = im
	}

	return yre*yre + yim*yim
}

File: ts/indicators/custom/goertzel-spectrum/estimator.ts
/** Internal Goertzel spectrum estimator. Port of MBST's GoertzelSpectrumEstimator used only
 * by the GoertzelSpectrum indicator; not exported from the module barrel. */
export class GoertzelSpectrumEstimator {
  public readonly length: number;
  public readonly spectrumResolution: number;
  public readonly lengthSpectrum: number;
  public readonly minPeriod: number;
  public readonly maxPeriod: number;
  public readonly isFirstOrder: boolean;
  public readonly isSpectralDilationCompensation: boolean;
  public readonly isAutomaticGainControl: boolean;
  public readonly automaticGainControlDecayFactor: number;

  public readonly inputSeries: number[];
  public readonly inputSeriesMinusMean: number[];
  public readonly spectrum: number[];
  public readonly period: number[];

  // Pre-computed trigonometric tables.
  private readonly frequencySin: number[]; // first-order only
  private readonly frequencyCos: number[]; // first-order only
  private readonly frequencyCos2: number[]; // second-order only

  public mean = 0;
  public spectrumMin = 0;
  public spectrumMax = 0;
  public previousSpectrumMax = 0;

  constructor(
    length: number,
    minPeriod: number,
    maxPeriod: number,
    spectrumResolution: number,
    isFirstOrder: boolean,
    isSpectralDilationCompensation: boolean,
    isAutomaticGainControl: boolean,
    automaticGainControlDecayFactor: number,
  ) {
    const twoPi = 2 * Math.PI;

    const lengthSpectrum = Math.trunc((maxPeriod - minPeriod) * spectrumResolution) + 1;

    this.length = length;
    this.spectrumResolution = spectrumResolution;
    this.lengthSpectrum = lengthSpectrum;
    this.minPeriod = minPeriod;
    this.maxPeriod = maxPeriod;
    this.isFirstOrder = isFirstOrder;
    this.isSpectralDilationCompensation = isSpectralDilationCompensation;
    this.isAutomaticGainControl = isAutomaticGainControl;
    this.automaticGainControlDecayFactor = automaticGainControlDecayFactor;

    this.inputSeries = new Array<number>(length).fill(0);
    this.inputSeriesMinusMean = new Array<number>(length).fill(0);
    this.spectrum = new Array<number>(lengthSpectrum).fill(0);
    this.period = new Array<number>(lengthSpectrum).fill(0);

    this.frequencySin = [];
    this.frequencyCos = [];
    this.frequencyCos2 = [];

    const result = spectrumResolution;

    if (isFirstOrder) {
      this.frequencySin = new Array<number>(lengthSpectrum).fill(0);
      this.frequencyCos = new Array<number>(lengthSpectrum).fill(0);

      for (let i = 0; i < lengthSpectrum; i++) {
        const period = maxPeriod - i / result;
        this.period[i] = period;
        const theta = twoPi / period;
        this.frequencySin[i] = Math.sin(theta);
        this.frequencyCos[i] = Math.cos(theta);
      }
    } else {
      this.frequencyCos2 = new Array<number>(lengthSpectrum).fill(0);

      for (let i = 0; i < lengthSpectrum; i++) {
        const period = maxPeriod - i / result;
        this.period[i] = period;
        this.frequencyCos2[i] = 2 * Math.cos(twoPi / period);
      }
    }
  }

  /** Fills mean, inputSeriesMinusMean, spectrum, spectrumMin, spectrumMax from the current
   * inputSeries contents. */
  public calculate(): void {
    // Subtract the mean from the input series.
    let mean = 0;
    for (let i = 0; i < this.length; i++) {
      mean += this.inputSeries[i];
    }
    mean /= this.length;

    for (let i = 0; i < this.length; i++) {
      this.inputSeriesMinusMean[i] = this.inputSeries[i] - mean;
    }
    this.mean = mean;

    // Seed with the first bin.
    let spectrum = this.goertzelEstimate(0);
    if (this.isSpectralDilationCompensation) {
      spectrum /= this.period[0];
    }

    this.spectrum[0] = spectrum;
    this.spectrumMin = spectrum;

    if (this.isAutomaticGainControl) {
      this.spectrumMax = this.automaticGainControlDecayFactor * this.previousSpectrumMax;
      if (this.spectrumMax < spectrum) {
        this.spectrumMax = spectrum;
      }
    } else {
      this.spectrumMax = spectrum;
    }

    for (let i = 1; i < this.lengthSpectrum; i++) {
      spectrum = this.goertzelEstimate(i);
      if (this.isSpectralDilationCompensation) {
        spectrum /= this.period[i];
      }

      this.spectrum[i] = spectrum;

      if (this.spectrumMax < spectrum) {
        this.spectrumMax = spectrum;
      } else if (this.spectrumMin > spectrum) {
        this.spectrumMin = spectrum;
      }
    }

    this.previousSpectrumMax = this.spectrumMax;
  }

  private goertzelEstimate(j: number): number {
    return this.isFirstOrder ? this.goertzelFirstOrderEstimate(j) : this.goertzelSecondOrderEstimate(j);
  }

  private goertzelSecondOrderEstimate(j: number): number {
    const cos2 = this.frequencyCos2[j];

    let s1 = 0;
    let s2 = 0;

    for (let i = 0; i < this.length; i++) {
      const s0 = this.inputSeriesMinusMean[i] + cos2 * s1 - s2;
      s2 = s1;
      s1 = s0;
    }

    const spectrum = s1 * s1 + s2 * s2 - cos2 * s1 * s2;
    return spectrum < 0 ? 0 : spectrum;
  }

  private goertzelFirstOrderEstimate(j: number): number {
    const cosTheta = this.frequencyCos[j];
    const sinTheta = this.frequencySin[j];

    let yre = 0;
    let yim = 0;

    for (let i = 0; i < this.length; i++) {
      const re = this.inputSeriesMinusMean[i] + cosTheta * yre - sinTheta * yim;
      const im = this.inputSeriesMinusMean[i] + cosTheta * yim + sinTheta * yre;
      yre = re;
      yim = im;
    }

    return yre * yre + yim * yim;
  }
}
